File: 101. Symmetric Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        p = root.left
        q = root.right
        return self.helper(p, q)
    
    def helper(self, p, q):
        if not p and not q:
            return True
        if (not p and q) or (not q and p):
            return False
        if p.val != q.val:
            return False
        return self.helper(p.left, q.right) and self.helper(p.right, q.left)

File: 101. Symmetric Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
    if (root === null)
        return true;
    var p = root.left;
    var q = root.right;
    return helper(p, q);
};

var helper = function(p, q)
{
    if (p === null && q === null)
        return true;
    if ((p !== null && q === null) || (p === null && q !== null))
    {
        return false;
    }
    if (p.val != q.val)
        return false;
    return helper(p.left, q.right) && helper(p.right, q.left);
};
